File: Data Structures & Algorithms/sliding-window-maximum/submission-1.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int left = 0;
        int right = 0;
        int[] result = new int[nums.length-k+1];

        int currentMax = -999999;
        while (right < nums.length) {
            while (right < k) {
                currentMax = Math.max(currentMax, nums[right]);
                right++;
            }
            result[right-k] = currentMax;
            if (right == nums.length) break;

            int leaving = nums[left];
            int joining = nums[right];

            if (leaving == currentMax) {
                int prevLeft = left;
                currentMax = -999999;
                while (left < right) {
                    left++;
                    currentMax = Math.max(currentMax, nums[left]);
                }
                left = prevLeft;
            }
            else {
                currentMax = Math.max(currentMax, joining);
            }

            left++;
            right++;
        }

        result[right-k] = currentMax;

        return result;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [1 for i in range(len(nums))]
        right = [1 for i in range(len(nums))]

        for i in range(1,len(nums)):
            if i == 1:
                left[i] = nums[i-1]
                right[len(nums)-i-1] = nums[len(nums)-i]
            else:
                left[i] = left[i-1] * nums[i-1]
                right[len(nums)-i-1] = right[len(nums)-i] * nums[len(nums)-i]
        
        result = [1 for i in range(len(nums))]

        for i in range(len(nums)):
            result[i] = left[i] * right[i]

        return result

        